### ISSUE TITLE
cpuset: failed allocations on nodes with free resources

### ISSUE THREAD

**Issue by jblernout** [hashicorp/nomad#18549]
### Nomad version
```
Nomad v1.6.2
BuildDate 2023-09-13T16:47:25Z
Revision 73e372ad94033db2ceaf53468b270a31544c23fd
```

### Operating system and Environment details
`Ubuntu 20.04`

### Issue

When running several service jobs in parallel and updating the task group count, the placement starts to fail at some point, returning core resource exhaustion while there are still free resources on the client.

Client Allocs:
![image](https://github.com/hashicorp/nomad/assets/87059/1579e5d4-fc21-4a53-a081-c563ff2cc49a)

Jobs:
![image](https://github.com/hashicorp/nomad/assets/87059/44d10353-efaa-4ef7-92a7-c70c432ac00e)

job0 (no issue)
![image](https://github.com/hashicorp/nomad/assets/87059/5733c106-bf1a-4e00-a15d-0764d2417c41)

job1 (placement failure)
![image](https://github.com/hashicorp/nomad/assets/87059/ab72c80f-0d5f-4e3b-887f-b2b3ec4f158c)

6 allocations are available on the client, 1 is consumed by job0, there should be 5 remaining for job 1 but it gets only 2 with a placement failure on the 3 remaining.

The jobs are pretty simple. They use the raw_exec driver, and launch a script which starts a process with a taskset on the cpuset provided by Nomad server in the NOMAD_CPU_CORES env var. The process displays the cpuset and consumes the CPU of this set.

When the issue appears, there is a discrepancy between the cpuset displayed by the process and the cpuset reported by Nomad server when listing the allocations of the node.

Here is the output of the 3 tasks running:
Allocation ba834c63:
![image](https://github.com/hashicorp/nomad/assets/87059/d97a2eaa-629c-497c-a878-eb0781da26f9)

Allocation 2e31255c:
![image](https://github.com/hashicorp/nomad/assets/87059/ab1bb132-675d-4378-b061-60c4048eeb6f)

Allocation bba5c077:
![image](https://github.com/hashicorp/nomad/assets/87059/b6ed3304-a66c-485c-b276-f6e021197b10)

And the output from the node allocations:
[allocations.txt](https://github.com/hashicorp/nomad/files/12684960/allocations.txt)

Two allocations are reporting the same reserved cores, but none match the cores reported by the process:
```
node XXX, alloc job1, job ba834c63-9067-9858-3bc5-7a8cbcb7a172: 4,5,6,7
node XXX, alloc job1, job 2e31255c-8765-2dbf-84e4-487f454a107a: 4,5,6,7
```

Once this issue occurs, adding more tasks to the node becomes impossible.

### Reproduction steps

- A Nomad server hosting 1 Client with 24 cores
- A script to start the process on the client (start.sh)
```
launch() {
  taskset -c $NOMAD_CPU_CORES /tmp/nono 2>&1 & vPid=$!

  wait $vPid
}


interruptTrap() {
  echo "$(date '+%Y-%m-%dT%T.%3N') Script interrupted"

  # Reset trap to avoid infinite loop
  trap " " SIGINT SIGTERM

  # Kill process
  if [ -n "$vPid" ]; then
    echo "$(date '+%Y-%m-%dT%T.%3N') Sending SIGTERM to pid: ${vPid}"
    kill -s SIGTERM ${vPid}
    wait ${vPid}
  fi

  # Kill process group processes (if some remain)
  echo "$(date '+%Y-%m-%dT%T.%3N') Sending SIGTERM to all process of the group"
  kill -s SIGTERM 0
  wait

  exit 0
}


# Intercept SIGINT (user interrupt CTRL+C), SIGTERM
trap interruptTrap SIGINT SIGTERM

launch
```
- A process to display the cores and consume the CPU (this is the process launched by the script)

```
package main

import (
	"fmt"
	"os"
	"runtime"
	"sync"
	"time"
)

//
func main() {
	fmt.Println("reserved nomad cpus :", os.Getenv("NOMAD_CPU_CORES"))
	fmt.Println("runtime.CPUS:", runtime.NumCPU())

	// waste some cpu
	FOO := 0
	wg := &sync.WaitGroup{}
	worker := func() {
		dt0 := time.Now()
		for time.Since(dt0) < time.Minute*15 {
			FOO++
			if FOO == 1000 {
				fmt.Println(FOO)
			}
		}
		wg.Done()
	}

	wg.Add(runtime.NumCPU())
	for i := 0; i < runtime.NumCPU(); i++ {
		go worker()
	}
	wg.Wait()
}
```

- A process to launch the jobs on the Nomad server and trigger the issue:

```
package main

import (
	"bytes"
	"fmt"
	"strings"
	"time"

	"github.com/hashicorp/nomad/api"
	"github.com/hashicorp/nomad/jobspec"
)

const hcl = `
job "hello" {
  datacenters = ["adv3g2"]
  type = "service"
  group "world" {
    count = 1
    task "script"{
      driver = "raw_exec"
      config {
        command = "/bin/bash"
        args = ["-c", "/tmp/start.sh"]
      }

      kill_signal="SIGTERM"
      kill_timeout="2m"

      resources {
        cores = 4
      }
    }
  }
}
`

//
func main() {
	conf := api.DefaultConfig()
	conf.Address = "https://XXX:4646"
	client, err := api.NewClient(conf)
	if err != nil {
		panic(err)
	}

	// Init jobs
	var jobs []*api.Job
	for i := 0; i < 2; i++ {

		job, err := jobspec.Parse(bytes.NewReader([]byte(hcl)))
		if err != nil {
			panic(err)
		}
		*job.Name = fmt.Sprintf("job%d", i)
		*job.ID = *job.Name

		jobs = append(jobs, job)
	}

	// Reset jobs count
	err = updateJobsCount(client, jobs, []int{0, 0})
	if err != nil {
		fmt.Printf("err: %w\n", err)
	}

	// Updates jobs count to trigger the issue
	err = updateJobsCount(client, jobs, []int{2, 2})
	if err != nil {
		fmt.Printf("err: %w\n", err)
	}

	err = updateJobsCount(client, jobs, []int{1, 5})
	if err != nil {
		fmt.Printf("err: %v\n", err)
		return
	}

}

func updateJobsCount(client *api.Client, jobs []*api.Job, counts []int) error {

	if len(jobs) != len(counts) {
		return fmt.Errorf("jobs and counts must have the same length")
	}

	// Apply the new count to each job
	for i, job := range jobs {
		err := updateJobCount(client, job, counts[i])
		if err != nil {
			return err
		}
	}

	err := checkNodes(client)
	if err != nil {
		return err
	}

	return nil
}

func updateJobCount(client *api.Client, job *api.Job, count int) error {

	fmt.Printf("Updating job %s count to %d\n", *job.Name, count)

	// Set the instances count!
	*job.TaskGroups[0].Count = count

	_, _, err := client.Jobs().Register(job, nil)
	if err != nil {
		return fmt.Errorf("register failed, job %s, count %d: %w", *job.Name, count, err)
	}

	// Wait a bit, but don't wait until the evaluation is done
	time.Sleep(5 * time.Second)

	return nil
}

func checkNodes(client *api.Client) error {
	nodes, _, err := client.Nodes().List(nil)
	if err != nil {
		return err
	}

	for _, node := range nodes {
		allocs, _, err := client.Nodes().Allocations(node.ID, nil)
		if err != nil {
			return err
		}

		sets := make(map[string]*api.Allocation)
		broken := false

		for _, alloc := range allocs {

			// Filter out old allocations and other jobs
			if _, ok := alloc.AllocatedResources.Tasks["script"]; !ok {
				continue
			}
			if *alloc.Job.Status != "running" || alloc.ClientStatus == "complete" || alloc.DesiredStatus == "stop" {
				continue
			}

			reservedCoresStr := getReservedCoresString(alloc)

			if savedAlloc, ok := sets[reservedCoresStr]; ok {
				fmt.Printf("Error, same reserved cores detected!\n")
				displayNodeAlloc(node, alloc)
				displayNodeAlloc(node, savedAlloc)

				broken = true
				break
			}

			sets[reservedCoresStr] = alloc

		}

		if broken {
			return fmt.Errorf("broken")
		}
	}

	return nil
}

func getReservedCoresString(alloc *api.Allocation) string {
	savedReservedCores := alloc.AllocatedResources.Tasks["script"].Cpu.ReservedCores
	return strings.Trim(strings.Join(strings.Fields(fmt.Sprint(savedReservedCores)), ","), "[]")
}

func displayNodeAlloc(node *api.NodeListStub, alloc *api.Allocation) {
	fmt.Printf("node %s, alloc %s, job %s: %s\n",
		node.Name,
		*alloc.Job.Name,
		alloc.ID,
		getReservedCoresString(alloc),
	)
}
```

However to access the `alloc.AllocatedResources.Tasks[...].Cpu.ReservedCores` field, the api package has been replaced with a custom version, where the `AllocatedCpuResources` struct has been modified to give access to the missing field

Update in the `go.mod` file:
```
module nomadissue

go 1.20

replace github.com/hashicorp/nomad/api => ./api

require (
	github.com/hashicorp/nomad v1.6.2
	github.com/hashicorp/nomad/api v0.0.0-20230913164725-73e372ad9403
)

require (
	github.com/gorilla/websocket v1.5.0 // indirect
	github.com/hashicorp/cronexpr v1.1.2 // indirect
	github.com/hashicorp/errwrap v1.1.0 // indirect
	github.com/hashicorp/go-cleanhttp v0.5.2 // indirect
	github.com/hashicorp/go-multierror v1.1.1 // indirect
	github.com/hashicorp/go-rootcerts v1.0.2 // indirect
	github.com/hashicorp/hcl v1.0.1-vault-5 // indirect
	github.com/mitchellh/go-homedir v1.1.0 // indirect
	github.com/mitchellh/mapstructure v1.5.0 // indirect
	golang.org/x/exp v0.0.0-20230905200255-921286631fa9 // indirect
)
```

And `go.work` file:
```
go 1.20

use (
	.
	api
)
```

`File api/allocations.go`:
```
type AllocatedCpuResources struct {
	CpuShares     int64
	ReservedCores []int64 // Missing ???
}
```

The process output should be the following:
```
Updating job job0 count to 0
Updating job job1 count to 0
Updating job job0 count to 2
Updating job job1 count to 2
Updating job job0 count to 1
Updating job job1 count to 5
Error, same reserved cores detected!
node XXX, alloc job1, job ba834c63-9067-9858-3bc5-7a8cbcb7a172: 4,5,6,7
node XXX, alloc job1, job 2e31255c-8765-2dbf-84e4-487f454a107a: 4,5,6,7
err: broken
```

#### Expected Result

The placement should not report exhausted core resources when the client has free resources available

#### Actual Result

The placement fails and reports exhausted core resources on the client while only half of its resources are being used

### Job file (if appropriate)

See source code above.

**Comment by lgfa29** [comment]
Thank you for the detailed report @jblernout.

It seems like the Nomad scheduler may be miscounting core assignments. 

> Once this issue occurs, adding more tasks to the node becomes impossible.

In this state, do you see any log messages in the servers that look like `plan for node rejected`?

**Comment by jblernout** [comment]
We didn't find any log message in the servers that looks like `plan for node rejected`

**Comment by lgfa29** [comment]
Ah ok, thanks for looking! 

I think parts of the code that touch this is being reworked, so we can check if this bug is fixed once that work is done.

**Comment by jblernout** [comment]
Do you know in which release this work should be done?

**Comment by shoenig** [comment]
Hi @jblernout the way Nomad manages CPU resources is being substantially overhauled for Nomad 1.7 (ETA November). However I'll also investigate this bug in the context of providing backport fixes for 1.6 and 1.5, but it's not the highest priority right now.

**Comment by jblernout** [comment]
Thanks for the update!

**Comment by jblernout** [comment]
Hi @shoenig,
I tested the 1.7.2 (server and client) and the issue persists :(

I also noticed another issue linked to this one:
Nomad often launches tasks (raw_exec) and assigns them the same cpuset (two tasks are launched with NOMAD_CPU_CORES containing the same set). It is problematic as some cpusets will be used for several tasks, and other cpusets won't be used. A workaround to this issue is ignoring the NOMAD_CPU_CORES env var and let the task resolve which cpusets are free and can be used.

Changes were introduced in the 1.7.x, where the cpuset chosen by Nomad is assigned to the cgroups of the task. Only the cpuset determined by Nomad can be used in the task. Trying to perform a taskset within the task with a different set will fail, with an "invalid argument" error. It prevents the workaround from working anymore.

**Comment by tgross** [comment]
Internal ref: https://hashicorp.atlassian.net/browse/NET-12140

**Comment by tgross** [comment]
My colleague @allisonlarson started working on this recently along with https://github.com/hashicorp/nomad/issues/25525 (which seems to be tightly related). The working reproduction has been very helpful. We'll update again when we have something more concrete to update on.

**Comment by allisonlarson** [comment]
Hi! I was able to reproduce this bug (using the reproduction steps), and have a PR up with the fix  https://github.com/hashicorp/nomad/pull/25705. Once its reviewed and merged, it'll get shipped in the next version of Nomad. Thanks for the detailed report!
